test: Run scripted-diff in subshell

dongcarl opened the pull request:
scripted-diffs should be run in subshells so that their execution does not
affect the shell variables of commit-script-check. Shell variables are not
unset before evaluating the scripted-diff, so that they might be used in
the subshell. To this end, the variable previously named i is now more
descriptively named commit, this also allows scripted-diffs to use the
commonly used variable i without fear of losing a reference to the
commit.

dongcarl commented:
Fixes concerns described in: https://github.com/bitcoin/bitcoin/pull/13743#pullrequestreview-166625569

practicalswift commented:
Concept ACK

Nice improvement to the scripted-diff integrity!

Would be nice and meta to have this change done as a `scripted-diff` :-)

practicalswift commented on test/lint/commit-script-check.sh:
```diff
@@ -20,23 +20,23 @@ fi
             echo "$SCRIPT"
-            eval "$SCRIPT"
-            git --no-pager diff --exit-code $i && echo "OK" || (echo "Failed"; false) || RET=1
+            (eval "$SCRIPT")
```
Would it make sense to use `(set -e; eval "$SCRIPT") || RET=1 && echo "scripted-diff execution failed."`?

That would allow us to catch execution failures within the scripted-diff run.

It probably shouldn't be done as part of this PR as that would change behaviour.

dongcarl replied:
Not sure about `set -e`, as I've found that people are not used to its behaviour and there are legitimate uses of not failing immediately upon the failure of one command. I think that as long as the diff produces the expected result, we should be good.

What I _would_ like to have is `set -x`, which doesn't change behaviour but allows us to see line by line execution fully expanded (in case it's a complex script they'd like to debug).

practicalswift replied:
`set -x` would be nice :-)

dongcarl committed:
scripted-diff: Run scripted-diff in subshell

-BEGIN VERIFY SCRIPT-
sed -i 's/\bi\b/commit/g' test/lint/commit-script-check.sh
sed -i '34s/eval "$SCRIPT"/(eval "$SCRIPT")/' test/lint/commit-script-check.sh
-END VERIFY SCRIPT-

dongcarl commented:
Made it `M E T A`

ryanofsky reviewed (approved):
utACK 43f909990190b3ff7883f0b2c117daa876d8fd99. I'd probably prefer not to set -x by default, and just leave this up to individual scripts, but I wouldn't oppose it. Maybe something to do in a followup PR.

Empact commented:
utACK 43f9099
